packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts

```diff
@@ -119,173 +122,181 @@ export default createESLintRule<Options, MessageIds>({
+      if (!sourceProject || !targetProject) {
+        return;
+      }
 
-        // project => npm package
-        if (isNpmProject(targetProject)) {
-          return;
-        }
-        // check constraints between libs and apps
-        // check for circular dependency
-        const circularPath = checkCircularPath(
-          projectGraph,
-          sourceProject,
-          targetProject
-        );
-        if (circularPath.length !== 0) {
-          context.report({
-            node,
-            messageId: 'noCircularDependencies',
-            data: {
-              sourceProjectName: sourceProject.name,
-              targetProjectName: targetProject.name,
-              path: circularPath.reduce(
-                (acc, v) => `${acc} -> ${v.name}`,
-                sourceProject.name
-              ),
-            },
-          });
-          return;
-        }
+      // same project => allow
+      if (sourceProject === targetProject) {
+        return;
+      }
 
-        // cannot import apps
-        if (targetProject.type === ProjectType.app) {
+      // project => npm package
+      if (isNpmProject(targetProject)) {
+        return;
+      }
+      // check constraints between libs and apps
+      // check for circular dependency
+      const circularPath = checkCircularPath(
+        projectGraph,
+        sourceProject,
+        targetProject
+      );
+      if (circularPath.length !== 0) {
+        context.report({
+          node,
+          messageId: 'noCircularDependencies',
+          data: {
+            sourceProjectName: sourceProject.name,
+            targetProjectName: targetProject.name,
+            path: circularPath.reduce(
+              (acc, v) => `${acc} -> ${v.name}`,
+              sourceProject.name
+            ),
+          },
+        });
+        return;
+      }
+
+      // cannot import apps
+      if (targetProject.type === ProjectType.app) {
+        context.report({
+          node,
+          messageId: 'noImportsOfApps',
+        });
+        return;
+      }
+
+      // cannot import e2e projects
+      if (targetProject.type === ProjectType.e2e) {
+        context.report({
+          node,
+          messageId: 'noImportsOfE2e',
+        });
+        return;
+      }
+
+      // buildable-lib is not allowed to import non-buildable-lib
+      if (
+        enforceBuildableLibDependency === true &&
+        sourceProject.type === ProjectType.lib &&
+        targetProject.type === ProjectType.lib
+      ) {
+        if (
+          hasBuildExecutor(sourceProject) &&
+          !hasBuildExecutor(targetProject)
+        ) {
           context.report({
             node,
-            messageId: 'noImportsOfApps',
+            messageId: 'noImportOfNonBuildableLibraries',
           });
           return;
         }
+      }
+
+      // if we import a library using loadChildren, we should not import it using es6imports
+      if (
+        node.type === AST_NODE_TYPES.ImportDeclaration &&
```

Comment:
I added this line because `TSESTree.ImportExpression` does not have `importKind`